API/ev3_timer.c

```diff
@@ -12,268 +16,552 @@
+    TimerExit();
+    return false;
 }
 
-unsigned long Timer(uint8_t Timer)
+void tmInitArrays()
 {
-  // slow timer 10 ticks per second
-  return FastTimer(Timer) / 10;
+    for (TimerInterval i = ti10ms; i < ti_count; i++)
+        tmArrayInit(&tcb.callbacksFor[i]);
 }
 
-unsigned long FastTimer(uint8_t Timer)
+bool tmInitKernel()
 {
-  // centisecond timer (100 ticks per second)
-  if (Timer < NUM_CS_TIMERS)
-	return TimerGetCS() - csTimers[Timer];
-  return 0;
+    struct epoll_event event = {0};
+    event.events = EPOLLIN;
+
+    // initialize epoll
+    tcb.waiterEpollFd = epoll_create(1);
+    RETURN_ON_FAIL(tcb.waiterEpollFd, "epoll_create1() failed", false);
+
+    for (TimerInterval i = ti10ms; i < ti_count; i++)
+    {
+        // create a timer object
+        int tfd = timerfd_create(CLOCK_MONOTONIC, 0);
+        RETURN_ON_FAIL(tfd, "timerfd_create() failed", false);
+
+        tcb.callbacksFor[i].timer_fd = tfd;
+        tcb.callbacksFor[i].enabled = false;
+
+        // and register it in the epoll object
+        event.data.u32 = i;
+        int result = epoll_ctl(tcb.waiterEpollFd, EPOLL_CTL_ADD, tfd, &event);
+        RETURN_ON_FAIL(result, "epoll_ctl(EPOLL_CTL_ADD) (timerfd registration) failed", false);
+    }
+
+    // initialize terminating eventfd
+    tcb.waiterTerminatorFd = eventfd(0, 0);
+    RETURN_ON_FAIL(tcb.waiterTerminatorFd, "eventfd() failed", false);
+
+    // and register it as well
+    event.data.u32 = TERMINATOR;
+    int result = epoll_ctl(tcb.waiterEpollFd, EPOLL_CTL_ADD, tcb.waiterTerminatorFd, &event);
+    RETURN_ON_FAIL(result, "epoll_ctl(EPOLL_CTL_ADD) (eventfd registration) failed", false);
+    return true;
 }
 
-void ClearTimer(uint8_t Timer)
+bool tmInitThread(void)
 {
-  if (Timer < NUM_CS_TIMERS)
-	csTimers[Timer] = TimerGetCS();
+    // create synchronization mutex
+    RETURN_ON_FAIL(pthread_mutex_init(&tcb.waiterMutex, NULL),
+                   "pthread_mutex_init(waiter) failed", false);
+
+    // create timestamp mutex
+    RETURN_ON_FAIL(pthread_mutex_init(&tcb.timestampMutex, NULL),
+                   "pthread_mutex_init(timestamp) failed", false);
+
+    // and start the thread
+    RETURN_ON_FAIL(pthread_create(&tcb.waiterThread, NULL, tmThreadEntry, NULL),
+                   "pthread_create() failed", false);
+    return true;
 }
 
-void SetTimer(uint8_t Timer, unsigned long Value)
+
+// THREAD LOOP //
+
+void *tmThreadEntry(void *arg)
 {
-  if (Timer < NUM_CS_TIMERS)
-	csTimers[Timer] = TimerGetCS() - Value;
+    (void) arg;
+
+    // initialize local shadow copy
+    // (used while calling handlers to prevent modification while iterating)
+    CallbackBlock localCopy;
+    tmArrayInit(&localCopy);
+
+    // block all signals to prevent interruptions & delays
+    tmThreadBlockSignals();
+
+    // until and error or interrupt happens
+    while (true)
+    {
+        struct epoll_event event = {0};
+
+        // try to read an event
+        if (!tmThreadWaitForEvent(&event))

```

Comment:
I'd use that as `while` condition.

Reply:
Yes, that should simplify the code. I will change it.

Reply:
Fixed in 7323f55428266a6226bc069dc88d68e524a64b6c :+1: